src/os/spawn.php

```diff
@@ -0,0 +1,173 @@
+<?hh
+
+namespace HH\Lib\Experimental\OS;
+use namespace HH\Lib\C;
+use namespace HH\Lib\Vec;
+use namespace HH\Lib\_Private\_OS;
+use type HH\Lib\OS\{FileDescriptor, IsNotADirectoryException};
+use function HH\Lib\OS\{close, open};
+
+newtype pid_t as int = int;
+newtype ExitCode as int = int;
+
+<<__Sealed(
+  posix_spawn_file_actions_addopen::class,
+  posix_spawn_file_actions_adddup2::class,
+  posix_spawn_file_actions_addchdir_np::class,
+)>>
+interface PosixSpawnFileActionsSetter {}
+
+final class posix_spawn_file_actions_addopen
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(
+    public int $filedes,
+    public string $path,
+    public int $oflag,
+    public int $mode,
+  ) {}
+}
+
+final class posix_spawn_file_actions_adddup2
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(
+    public FileDescriptor $filedes,
+    public int $newfiledes,
+  ) {}
+}
+
+final class posix_spawn_file_actions_addchdir_np
+  implements PosixSpawnFileActionsSetter {
+  public function __construct(public string $path) {
+  }
+}
+
+type posix_spawn_file_actions_t = vec<PosixSpawnFileActionsSetter>;
+
+const int POSIX_SPAWN_SETPGROUP = 2;
+const int POSIX_SPAWN_SETSID = 128;
+
+type PosixSpawnFlags = int;
+
+type posix_spawnattr_t = shape(
+  ?'posix_spawnattr_setpgroup' => pid_t,
+  ?'posix_spawnattr_setflags' => PosixSpawnFlags,
+);
+
+function posix_spawnp(
+  string $file,
+  posix_spawn_file_actions_t $file_actions,
+  posix_spawnattr_t $attributes,
+  vec<string> $argv,
+  vec<string> $envp,
```

Comment:
I don't have strong feelings here, but needs asking: should this be a `dict<string, string>` instead?

- `vec` is closer to the C
- `dict` makes more sense for Hack in general, but would raise issues like 'what does = in a variable name mean?'

Reply:
I think it would make sense to throw if `=` or `\0` are present either in the key or value ( https://github.com/azjezz/psl/blob/2.0.x/src/Psl/Env/get_var.php#L21 )

Reply:
I think we would better treat `\0` the same as other `HH\Lib\OS` functions. Do they truncate anything after `\0` silently or throw an error for `\0`?

Reply:
for the most part, pass verbatim to the C library, which will *usually* truncate - except in cases where the C library takes both a pointer and length parameter.

Reply:
Why not make the type opaque and use vec-string under the hood? The calling code would look like.

```HACK
Process\posix_spawnp(
  ...,
  Process\envp_from_dict(dict['K' => 'V']),
);

Process\posix_spawnp(
  ...,
  Process\envp_from_vec(vec['K=V']),
);
```

envp_from_dict would check if the key does contain `=`.

The argument should have a default value.

Reply:
The PR title is a bit inaccurate - this is just for `OS\`. I like the idea for `Process\`, but feels like it's a higher level abstraction than belongs in `OS\`

Reply:
I see. It should be a vec nevertheless, since having "duplicate keys" is something you _could_ need.